modules/puppet/templates/govuk_puppet

```diff
@@ -8,11 +8,200 @@ fi
+remove_all_locks() {
 
-TIME_TOOK=$((`date +%s%3N`-START_TIME))
-echo "<%= @fqdn_metrics %>.puppet.run_duration:${TIME_TOOK}|ms" | nc -w 1 -u localhost 8125
+	if ls *.lock 1> /dev/null 2>&1; then #there are locks
 
-# We should only reach here if a last_run_summary was written.
-/usr/local/bin/puppet_passive_check_update >/dev/null
+                        for filename in *.lock
+                        do
+                                rm $filename
+                                echo "Removed: $filename"
+                        done
+	fi
+
+	shift #Pop off --remove-all-locks
+
+	do_puppet --enable "$@" #Re-enable puppet
+
+}
+
+listlocks() {
+
+	#list all current locks, date, name, reason, identifier
+        #simplify this structure. push the date and a (kind of) 
+	#uuid to the filename. Keep the reason, name and uuid as
+	#the lockfile content
+
+	#NB The last field (the lock identifier) can be grabbed with
+	#a simple awk '{print $NF}' in a script
+
+	if ls *.lock 1> /dev/null 2>&1; then
+
+        	for filename in *.lock
+               	 	do
+               	        	line=$(head -n 1 "$filename") #Grab file contents
+               	         	echo "$filename : $line"
+               	 	done
+
+	else
+		echo "No lock files present"
+	fi
+
+}
+
+do_puppet() {
+
+	# Allow return code 2 in case -v/--detailed-exitcodes is passed.
+	RBENV_VERSION=1.9.3 /usr/local/bin/govuk_setenv default puppet agent --onetime --no-daemonize "$@" || [ $? -eq 2 ]
+
+	TIME_TOOK=$((`date +%s%3N`-START_TIME))
+	echo "<%= @fqdn_metrics %>.puppet.run_duration:${TIME_TOOK}|ms" | nc -w 1 -u localhost 8125
+
+	# We should only reach here if a last_run_summary was written.
+	/usr/local/bin/puppet_passive_check_update >/dev/null
+	exit 0;
+
+}
+
+
+usage() {
+
+	programname=$0
+
+	echo "usage: $programname"
+	echo "	--disable : add a lock; returns a numeric identifier"
+	echo "	--enable : remove a single lock; takes numeric identifier as parameter"
+	echo "	--list-locks : show all current locks"
+	echo "	--remove-all-locks : remove all locks and enable govuk_puppet"
+
+	exit 1
+
+}
+
+GOVUK_STATE_DIR="/var/run/lock/puppet" #created by puppet
+
+#change working directory or die
+cd "$GOVUK_STATE_DIR" || { echo "Unable to access locks directory: '$GOVUK_STATE_DIR'" 1>&2 ; exit 1; }
+
+OPTION=${1:-}
+
+if  [ -z "$OPTION" ]
+then
+	OPTION=help #Show usage if no command line arguments are present
+fi
+
+case "$OPTION" in
+
+--disable)
+
+	#require a reason, determine a name if interactive
+	#append a datetime stamp to assist in locating old / forgotten locks
+	#write the lockfile containing the reason and name, then 
+	#do puppet agent --disable
+
+	LOCK_DATE=`date +%Y-%m-%d`	
+
+	REASON=${2:-} 
+
+	if [ -z "$REASON" ] #Check if a reason has been passed
+	then 
+		#It wasn't
+		echo "Reason required as second parameter (quote-wrap); exiting."
+		exit 1;
+	else
```

Comment:
I don't think you need this code. If $2 was set then $REASON already contains it.